### PR TITLE
Fix problem where a fast client can "drown" a server

### PR DESCRIPTION
**Pull request by dvdplm**
When a `yamux::Connection` is setting up a new incoming connection ([`process_incoming()`](https://github.com/paritytech/yamux/blob/master/src/connection.rs#L146)) it tries to make as much progress as possible. 
If a client connection is fast enough to send all its data and then closes the connection, the `Connection` reads all the data within a single call to `process_incoming()`, including the end of the stream. This means that [it will set `is_dead` to true](https://github.com/paritytech/yamux/blob/master/src/connection.rs#L337) on the connection and clear the stream, which will cause `poll()` to [return `Ready(None)`](https://github.com/paritytech/yamux/blob/master/src/connection.rs#L148) which bubbles up to the muxer where it is interpreted as if the connection was gone. The result is that `muxing::inbound_from_ref_and_wrap()` [returns `None`](https://github.com/libp2p/rust-libp2p/blob/master/core/src/muxing.rs#L176) and the data read is lost.
This PR re-orders things such that `poll()` first checks for incoming connections and then check for a dead connection and clears the stream if needed.

### CODE REVIEW

File: src/connection.rs [hunk 1]
```diff
@@ -497,6 +499,7 @@ where
 }
 
 /// A handle to a multiplexed stream.
+#[derive(Debug)]
```
Comment (dvdplm) [inline src/connection.rs]: This has nothing to do with the PR subject. Ok to keep it here anyway?

File: src/connection.rs [hunk 2]
```diff
@@ -159,6 +156,12 @@ where
             connection.on_drop(Action::None);
             return Ok(Async::Ready(Some(stream)))
         }
+
+        if connection.is_dead {
+            connection.streams.clear();
```
Comment (dvdplm) [inline src/connection.rs]: Should the `notify_all()` call be here or where it was originally?

**Comment by twittner** [comment]
I agree that short lived connections that come and go within a single `poll` call may not produce any incoming streams. To handle this properly I have made some more changes, which are the subject of PR #26.

File: src/connection.rs (hunk 1 above)
Comment (dvdplm) [reply src/connection.rs]: I guess this is the only thing added by this PR! :D

Comment (twittner) [reply src/connection.rs]: Forgot about this one ;-). Have added it to #26.

**Comment by dvdplm** [comment]
Closing in favour of #26

**Pull request closed without merging**